### ISSUE TITLE
Remove virtual columns from the spec

### ISSUE THREAD

**Issue by rufuspollock** [w3c/csvw#718]
- virtual columns are clearly a processing step and are likely relatively complex. Adding these to the spec brings substantial additional complexity with limited benefit.
- Already non-normative in the spec.

Removal will simplify without losing meaningful features.

**Comment by gkellogg** [comment]
The ability to have multiple subjects per row is a valuable feature, and virtual columns are necessary to add the properties which connect these together.

For example, if I have "name" and "company" columns, I may indicate that the "name" is that of a person, and "company" the name of a company she works for. This requires adding data to get something like the following:

```
{
  "@type": "Person",
  "name": "Gregg Kellogg",
  "affiliation": {
    "@type": "Company",
    "name": "Kellogg Associates"
  }
}
```

In terms of complexity, not really, other than for validation. They're just treated as extra blank columns.

**Comment by JeniT** [comment]
-1

Agree with @gkellogg's points: virtual columns add significantly to what you can do when transforming into JSON or RDF. Also just to correct: virtual columns are normative, not non-normative.

**Comment by rufuspollock** [comment]
Again metadata spec is getting info that could - and should IMO - be in a separate transformation spec - cf #722.

To be clear, i get this is a use case I'm just dubious that adding this kind of programmatic conversion info into a metadata spec for CSV justifies the complexity.

The challenge here is that there are many features which add some value, but collectively we end up with a complex spec that is hard to use and implement. (To take an example: should processors always respect virtual columns, what happens if a virtual column has the name of another column etc etc).

I guess I should also ask out loud: is conversion of CSV to RDF a first-class use case of the metadata spec? For me it is not, but obviously if people think it is that would help clarify where we are all coming from.

Right re non-normative (I mis-read there!).

**Comment by danbri** [comment]
-1

This functionality (which allows CSV to substitute for Microdata/RDFa/JSON-LD for data feeds) is very important from Google's perspective. The group agonized carefully over expressivity here, and selected a design that captures a _lot_ of real world structured data, while providing extensibility hooks for more ambitious efforts (e.g. using R2RML, as mentioned in charter).

**Comment by rufuspollock** [comment]
@danbri but what about having this kind of transformation stuff in a separate mini-spec (already this family contains several different mini-specs) as per #722 - if transformation is really important (as it appears to be) let's address it properly and in its own mini-spec that is part of this family (and that makes it easier for users to decide what parts they want - many users may not want the expressivity that google needs for example).

**Comment by danbri** [comment]
If W3C thought that only Google cared about the RDF mapping, it would not have chartered and scoped the work to include it - http://www.w3.org/2013/05/lcsv-charter#scope lists the RDF mapping as one of the use cases for this metadata. Note that we also had review comments from @chaals at Yandex back in June (#552); the Web search use cases extend naturally to any service that is using schema.org. 

That said it may be that you have identified an opportunity for editorial rather than substantive changes to the specification(s), i.e. the richer facilities shouldn't clutter the presentation of the basics within our docs.  Specific suggestions on wording / polish there could be very useful.

Note that RDF is not the only use for virtual columns - http://www.w3.org/TR/2015/CR-tabular-metadata-20150716/#use-of-virtual-columns shows how it can help generate more idiomatic JSON.

**Comment by rufuspollock** [comment]
@danbri my question here was about it in the _metadata_ spec vs e.g. separate transformation spec (#722) _and_ its priority vs other items

BTW re #552 I've re-read and still struggling to understand the request - is the best summary your comment in https://github.com/w3c/csvw/issues/552#issuecomment-101706006?

Anyway, I can appreciate that virtual columns are a big item and if people feel should be in metadata rather than transformation spec then I get that :-)

**Comment by danbri** [comment]
The point about #552 was just that the use cases we care about at Google are also of interest to other search engines, as evidenced by @chaals review of the specs.

The trouble with partitioning out "transformations" (which we already do, see the generating RDF and generating JSON specs) is that those activities have impact (a.) on the tabular data model (b.) on the metadata. Having the entire data model specified in one place, and the entire metadata vocabulary specified in another, is also a kind of simplicity. Having watched the spec editors team try various partitionings, move  text around, tweak dependencies and wording, and respond carefully to previous issues on this, I trust them to have found a reasonably sensible balance here. No partitioning will ever be perfect but I have no doubt that the choices they've made have taken simplicity into account throughout. 

You might choose to argue that the contents and meaning of the metadata file should be defined in several documents (eg. core, transformations) rather than one, and that we should also have a "core datamodel" and "augmented for transformations datamodel" defined in two different specifications. And that to do would be simpler rather than more complex. But the time for arguing that (questionable...) position has long past, hence my -1.

**Comment by 6a6d74** [comment]
-1 to Rufus' proposal.

I agree with the comments from Jeni, Dan and Gregg. For me, transformation
of CSV to a form (RDF or [plain-old] JSON) that can be used in machine
processing steps is a primary use case. IMO, having "all the data model"
and "all the metadata vocabulary" in a single place makes for simplicity. I
can see the point that virtual columns add to implementation complexity;
but users of CSVW can be completely ignorant of the capability and hence
retain a simple approach (what they don't need, they don't need to know).

Jeremy

**Comment by rufuspollock** [comment]
@danbri i was originally one of the editors - and was always an advocate of the simpler approach (and not including transformations) as @JeniT will testify ;-)

My point, to reiterate, is that transformations are complex, not totally specified at present (IMO) and add a lot of implementational complexity. I understand re time though I have fairly consistently advocated this position - though not always been present sufficiently to argue the case!

@6a6d74 I think we probably just have different goals: transformation to RDF has never been an important use case for me and I do not think it is one for the vast majority of the potential users of the spec (but i could be wrong). I note that we already have the data model and metadata spec separate - what i'm suggesting is that transformation info including e.g. virtual columns could be in a separate mini-spec.

tbh though the virtual columns is relatively minor and happy to see it go either way.

**Comment by pmika** [comment]
-1 

As shown by the example of @gkellogg and the one in the spec, this functionality is crucial for mapping tabular data to more complex data structures, and required, for example, for mapping highly normalized tabular data typically found on the Web to more entity-focused data structures typically found in Knowledge Graphs/Knowledge Bases, such as those we maintain at Yahoo. This need is not specific to RDF (which is just one type of E/R model) but generally required for mapping structured data from one representation to another.

**Comment by iherman** [comment]
Just want to record my -1. 

This feature was actually the subject of very long discussions in the group, and the complexity issue was one of the issues that made this discussion so long and tedious. However, we eventually made a decision and I do not see any new argument that would warrant reopening this decision.

B.t.w., virtual columns are not only for RDF. The [example in the current JSON transformation spec](http://w3c.github.io/csvw/publishing-snapshots/CR-csv2json/Overview.html#h-example-events-listing) shows a usage for a real-life example that was collected for the use case and requirements document.

**Comment by rufuspollock** [comment]
@iherman i agree this is not RDF only - that was a comment to jtandy on his particular point.

I know this was a long discussion - so long, i actually got lost (though i did still record my doubts - perhaps not forcefully enough - at the time).

@pmika this sounds interesting - could you explain to me a bit more how you plan to use this. I've done a lot of mapping of tabular data into other databases and I'm still not quite getting why this feature is essential vs having this in your ETL system. My precise concern here is having this spec having some incomplete "ETL" style features in them is we get the worst of both worlds then: an overly complex spec but one which does not actually do much useful ETL.

I do get the goldilocks principle - sometimes a bit of compromise may be useful and this may be such an occasion - but i would be _really_ interested to understand your precise use case in detail (e.g. could you explain what kind of mappings you are anticipating doing from an exmpale CSV file, and what is your target DB, what you use the data for at yahoo etc). If this is already detailed in another issue do let me know ;-)

Aside: I don't think our goal here is to make CSV + our metadata as rich as e.g. RDF for describing data (if we wanted that we could use RDF ...)

**Comment by pmika** [comment]
@rgrp You are right that these are ETL features that may be implemented using other transformation mechanisms. In my practical experience though, there are lots of benefits to doing ETL close to the source, i.e. close to the people and systems who understand the semantics of the data.  You are also right that ETL can get arbitrarily complex to the point where it can't be (easily) done declaratively any more, but I would rather have the authors/publishers of the table being able to provide semantics to their data, rather than someone else who doesn't know the data that well having to recreate it. 

I'm not exactly 100% sure what the charter of this group, since I haven't followed the work that closely, but I'm hoping it will result in mechanisms to make data tables on the Web (CSV, HTML...) useful for intelligent applications.

**Comment by rufuspollock** [comment]
@pmika ok though i think we are being optimistic about _publishers / authors_ adding this: they have minimal incentive since they are not doing the ETL - you are :-)

That would be a fairly general critique: IME you have to be pretty realistic about what _publishers_ will adopt and use. The more complex we make this the less likely they are to use it. Sure we'd like them to add a bunch of useful metadata but will they? I've extensive personal experience with gov and others and that experience suggest it is tough to get even reasonably formed CSV let alone nice metadata - as one example see http://okfnlabs.org/bad-data/ex/gla-spending/. As we add complexity to the spec we increase the chance that publishers won't use it or will use it wrong ...

**Comment by pmika** [comment]
@rgrp This comes pay-as-you-go in the current form, and shouldn't cause significant additional effort for publishers not using these features. 

At the same time, not having these features would mean that this is just a standardization of CSV. Nice to have, but doesn't do much for moving ahead the current state-of-affairs for structured data on the Web, and hence not particularly interesting for large consumers of Web content, including search engines such as Yahoo.

**Comment by rufuspollock** [comment]
@pmika I'd argue the major challenge right now actually _is_ some basic improvement on the quality of CSV ;-) (given my experience of what is out there).

And the challenge is that without publishes adopting it won't matter what consumers _want_ (I've wanted way better csv from publishers from years but i haven't necessarily got it ;-) ...)

**Comment by pmika** [comment]
@rgrp I would leave the group to discuss further but I would get confirmation from W3C and its members whether this proposed rec is relevant to this particular consortium if it doesn't contribute to web architecture.

**Comment by JeniT** [comment]
Closed in favour of editorial reworking ref #722
